remote-config/src/fetch/fetcher.rs

```diff
@@ -0,0 +1,316 @@
+    pub fn new(invariants: ConfigInvariants) -> Self {
+        ConfigFetcherState {
+            target_files_by_path: Default::default(),
+            endpoint: get_product_endpoint(PROD_INTAKE_SUBDOMAIN, &invariants.endpoint),
+            invariants,
+            expire_unused_files: true,
+        }
+    }
+
+    /// To remove unused remote files manually. Must not be called when auto expiration is active.
+    /// Note: careful attention must be paid when using this API in order to not deadlock:
+    /// - This files_lock() must always be called prior to locking any data structure locked within
+    ///   FileStorage::store().
+    /// - Also, files_lock() must not be called from within FileStorage::store().
+    pub fn files_lock(&self) -> ConfigFetcherFilesLock<S> {
+        assert!(!self.expire_unused_files);
+        ConfigFetcherFilesLock {
+            inner: self.target_files_by_path.lock().unwrap()
+        }
+    }
+}
+
+pub struct ConfigFetcher<S: FileStorage> {
+    pub file_storage: S,
+    state: Arc<ConfigFetcherState<S::StoredFile>>,
+    timeout: AtomicU32,
+    /// Collected interval. May be zero if not provided by the remote config server or fetched yet.
+    pub interval: AtomicU32,
+}
+
+#[derive(Default)]
+pub struct OpaqueState {
+    client_state: Vec<u8>,
+}
+
+impl<S: FileStorage> ConfigFetcher<S> {
+    pub fn new(file_storage: S, state: Arc<ConfigFetcherState<S::StoredFile>>) -> Self {
+        ConfigFetcher {
+            file_storage,
+            state,
+            timeout: AtomicU32::new(5000),
+            interval: AtomicU32::new(0),
+        }
+    }
+
+    /// Quite generic fetching implementation:
+    ///  - runs a request against the Remote Config Server,
+    ///  - validates the data,
+    ///  - removes unused files
+    ///  - checks if the files are already known,
+    ///  - stores new files,
+    ///  - returns all currently active files.
+    /// It also makes sure that old files are dropped before new files are inserted.
+    pub async fn fetch_once(
+        &mut self,
+        runtime_id: &str,
+        target: Arc<Target>,
+        config_id: &str,
+        last_error: Option<String>,
+        opaque_state: &mut OpaqueState,
+    ) -> anyhow::Result<Vec<Arc<S::StoredFile>>> {
+        let Target { service, env, app_version } = (*target).clone();
+
+        let mut cached_target_files = vec![];
+        let mut config_states = vec![];
+
+        for StoredTargetFile { state, meta, .. } in self.state.target_files_by_path.lock().unwrap().values() {
+            config_states.push(state.clone());
+            cached_target_files.push(meta.clone());
+        }
+
+        let config_req = ClientGetConfigsRequest {
+            client: Some(datadog_trace_protobuf::remoteconfig::Client {
+                state: Some(ClientState {
+                    root_version: 1,
+                    targets_version: 0,
+                    config_states,
+                    has_error: last_error.is_some(),
```

Comment:
What about a `&mut Option<String>` and the code calls `take` in here? It will reduce the risk of receiving the same error multiple times.

Reply:
I would say that it's more idiomatic rust to rely on the fact that Option<String> is not a reference at all, i.e. is consumed on call.